src/Compilers/CSharp/Portable/Binder/CancellationTokenLocalScopeBinder.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// This binder is responsible for introducing the 'cancellationToken' variable
+    /// in async-iterator methods.
+    /// </summary>
+    internal sealed class CancellationTokenLocalScopeBinder : LocalScopeBinder
+    {
+        private readonly SourceMethodSymbol _methodSymbol;
+
+        public CancellationTokenLocalScopeBinder(SourceMethodSymbol owner, Binder enclosing)
+            : base(enclosing, enclosing.Flags & ~BinderFlags.AllClearedAtExecutableCodeBoundary)
+        {
+            Debug.Assert(!enclosing.Flags.Includes(BinderFlags.InCatchFilter));
+            Debug.Assert((object)owner != null);
+            Debug.Assert(enclosing is InMethodBinder);
+            _methodSymbol = owner;
+        }
+
+        protected override SourceLocalSymbol LookupLocal(SyntaxToken nameToken)
+            => throw ExceptionUtilities.Unreachable;
```

Comment:
Why throw here vs. calling through `Next.`?

Reply:
I looked at the callers of this method and none can presently reach here. Calling to `Next` seems fine though (can't test now, but may prevent surprises in the future)

Reply:
The one I saw that could conceivably reach here was deconstruction. Binders is a weak area for me though. It's possible I'm just mis-understanding the contract here.

---
In reply to: [271585392](https://github.com/dotnet/roslyn/pull/34692#discussion_r271585392) [](ancestors = 271585392)